11 glass pipelines

bishax opened the pull request:
Addresses #11

bishax committed:
WIP work on rewriting glass pipelines w/ ORMs

bishax committed:
Formatting

bishax committed:
ORM WIP

bishax committed:
orm_playground file

bishax committed:
Delete scrap file & fix s3fs dev mount

bishax committed:
Add association obj pattern to organisation_address

Plus extra comments/examples

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+    addresses = relationship("OrganisationAddress", back_populates="organisation")
+    # addresses = relationship(
+    #     "Address", secondary=organisation_address, back_populates="organisations"
+    # )
```
Do you also want to add in a `relationship` to `OrganisationCompaniesHouseMatch` here?

jaklinger replied:
Do you also want to add in a `relationship` to `OrganisationCompaniesHouseMatch` here?

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+class OrganisationAddress(Base):
+    """Association object between organisations and addresses"""
+
+    org_id = Column(Integer, ForeignKey("organisation.org_id"), primary_key=True)
```
Make sure to set `autoincrement=False` else prepare yourself for a world of pain

jaklinger replied:
(do this for all pks on Integer columns in this ORM, I've added this to the list of boilerplate stuff)

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+
+    org_id = Column(Integer, ForeignKey("organisation.org_id"), primary_key=True)
+    address_id = Column(Integer, ForeignKey("address.address_id"), primary_key=True)
+    # TODO: do we really want to keep ranks? If ranks constantly change then
```
regarding this TODO: is the intention that you a) upsert in new collections or b) create duplicate entries in order to create a longitudinal dataset? If the former, then the ranks changing doesn't make a difference to us. If the latter, then you will also need to add `date` to the primary key (but keep the index as well), and then the proliferation of data shouldn't make too much difference since slicing on the date will be quick (unless you're worried about diskspace, but don't)

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+    rank = Column(
+        Integer, nullable=False, index=True, doc="Rank of `address_id` for `org_id`"
+    )
+    is_active = Column(
```
Who creates the `is_active` field: you or glass? If it's them, then this fits with taking longitudinal data, but if it's you then I'm assuming you're going for upserts

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+        doc="Indicates whether `address_id` is active",
+    )
+    # TODO : should this really have two date columns: valid_from and valid_to ?
+    date = Column(
```
perhaps best to change to `date_insertion` and `date_updated`. I know you said you were worried about polluting columns with boilerplate, but you could just use a mixin with these fields to add them into your ORMs automagically

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+#     # TODO: Date Date column indicating date of latest dump with this relationship
+# )
+class OrganisationAddress(Base):
+    """Association object between organisations and addresses"""
```
for all tables, I would set the `__tablename__` class variable so that the table names are prefixed with something like `glass_`

bishax replied:
Ok, I'd abstracted the boilerplate of setting of `__tablename__` away here: https://github.com/nestauk/research_daps/pull/13/files/54ff408f04d3f8591fb2563463d04b13c54d3bf8#diff-554965363b010d5c38b61d0ffc80e72888e90ff533731b41fe69cc990059a0ff

I could extend this to take a prefix.

jaklinger replied:
ah I see, sorry! i didn't see that `Base` was from `research_daps` as I implemented in `ojd_daps`. In the case of `ojd_daps` table prefixes aren't necessary since there is only one project, whereas I guess for `research_daps` there will be multiple projects and so there will be ambiguity that can be cleared up by project-specfici prefixes

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+        doc="Organisation name inferred by named entity recognition",
+        index=True,
+    )
+    website = Column(String, nullable=False, doc="URL of organisation")
```
No strong feeling, but I opt for `TEXT` over `String` because it's more explicit (`String` will render to `TEXT` anyway, and `String(123)` will render to something like `TINYTEXT`, although I think reading at a glance you would think this was `VARCHAR(123)`)

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+    has_webshop = Column(
+        Boolean, nullable=False, doc="If True, presence of a webshop was found"
+    )
+    vat_number = Column(String, doc="VAT number (low population)")
```
what does `(low population)` mean?

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+    )
+
+
+class OrganisationDescription(Base):
```
Is the idea here that a) Glass already provide multiple descriptions or b) You are intending to update with multiple descriptions over time? If b) I think the up{sert/dat}ing logic isn't consistent across the ORMs but if a) then fine!

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+    """List of addresses found in websites"""
+
+    address_id = Column(Integer, primary_key=True)
+    address_text = Column(String, nullable=False, unique=True, doc="Full address text")
```
The `unique` check on this is quite heavy, and unlikely to catch many exceptions (and even if they do, it probably won't change much?)

jaklinger replied:
or is this to trigger the autopk? either way, on ingestion it could get quite heavy then needing to generate the link table back to orgs

jaklinger commented on research_daps/orms/glass.py:
```diff
@@ -0,0 +1,202 @@
+class Address(Base):
+    """List of addresses found in websites"""
+
+    address_id = Column(Integer, primary_key=True)
```
Who defines an address id (you or glass?) How can it be defined to be unique? What if there is a subtle change in address (e.g. `5 Church Road, Office 2` to `Office 2, 5 Church Road`)?

bishax committed:
WIP - horrendous mono commit to be factored out

bishax committed:
companies house dump use later zip version

bishax committed:
mysql config

bishax committed:
WIP flow pipelines

- CH processing
- Glass processing
- Glass temp merge flow (to be taken over by luigi tasks)
- Glass notice processing
- Glass notice temp merge flow (to be taken over by luigi tasks)

bishax committed:
Dependencies for research_daps

bishax committed:
WIP luigi tasks for glass, glass notices, Companies house

Lots of performance issues to iron out

bishax committed:
Deleted dev script

bishax committed:
.gitignore update

bishax committed:
Flow to merge companies house flows

bishax committed:
fill nan address text & includes postcode; filter "empty" addresses

jaklinger commented:
@bishax is this good to close? Just cleaning up the ol' inbox before I leave and I can see that this is dangling

bishax committed:
Merge branch 'dev' into 11_glass_pipelines
